sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala

```diff
@@ -293,14 +295,40 @@ private[cosmos] class CosmosRowConverter(
         case DecimalType() =>
           convertToJsonNodeConditionally(rowData.asInstanceOf[java.math.BigDecimal])
         case DateType if rowData.isInstanceOf[java.lang.Long] =>
-          convertToJsonNodeConditionally(rowData.asInstanceOf[Long])
+          serializationConfig.serializationDateTimeConversionMode match {
+            case SerializationDateTimeConversionModes.Default =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[Long])
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(LocalDate
+                .ofEpochDay(rowData.asInstanceOf[Long])
+                .atStartOfDay()
+                .toInstant(ZoneOffset.UTC).toEpochMilli)
+          }
         case DateType if rowData.isInstanceOf[java.lang.Integer] =>
-          convertToJsonNodeConditionally(rowData.asInstanceOf[Integer])
+          serializationConfig.serializationDateTimeConversionMode match {
+            case SerializationDateTimeConversionModes.Default =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[java.lang.Integer])
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(LocalDate
+                .ofEpochDay(rowData.asInstanceOf[java.lang.Integer].longValue())
+                .atStartOfDay()
+                .toInstant(ZoneOffset.UTC).toEpochMilli)
+          }
         case DateType => convertToJsonNodeConditionally(rowData.asInstanceOf[Date].getTime)
         case TimestampType if rowData.isInstanceOf[java.lang.Long] =>
-          convertToJsonNodeConditionally(rowData.asInstanceOf[Long])
+          serializationConfig.serializationDateTimeConversionMode match {
+            case SerializationDateTimeConversionModes.Default =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[Long])
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[Long].asInstanceOf[Instant].toEpochMilli)
+          }
         case TimestampType if rowData.isInstanceOf[java.lang.Integer] =>
-          convertToJsonNodeConditionally(rowData.asInstanceOf[Integer])
+          serializationConfig.serializationDateTimeConversionMode match {
+            case SerializationDateTimeConversionModes.Default =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[java.lang.Integer])
+            case SerializationDateTimeConversionModes.AlwaysEpochMilliseconds =>
+              convertToJsonNodeConditionally(rowData.asInstanceOf[java.lang.Integer].asInstanceOf[Instant].toEpochMilli)
```

Comment:
can we cast from Integer to Instant directly?

Reply:
No - but this conversion would have thrown ClassCastException anyway - missed to port the fix from convertSparkDataTypeToJosnNodeNonNull